### PR TITLE
Use netCDF4 as the NetCDF dependency instead of Scientific.IO/NetCDF

### PR DESCRIPTION
**Pull request by eurydice76**
In the process of getting rid of ScientificPython and MMTK, I use in this PR a dependency to NetCDF4 library which is standard (can you please install it in the build machine). I had to change the unit test and now some are breaking again. Could you please investigate on this Rastislav as you were on that topic a few days ago ? Many thanks in advances.

### CODE REVIEW

**Comment by eurydice76** [comment]
I feel sorry for having asked you to solve the failing build. I did not think it would be such a fight. Could you tell me what is going on ? Is that the adding of NetCDF4 library which makes the things so hard ?

**Comment by RastislavTuranyi** [comment]
That's just how things are with the pipeline, since I have to push every time I want to test the changes I made. I have solved the issue with Ubuntu CI/CD, where it turned out that I had to alter the `PATH` and `LD_LIBRARY_PATH` environmental variables before running the tests so that the location where apt installs netcdf and libnetcdf is searched (for some reason it was not necessary before).

The only thing that remains is this bizarre error generated by `BuildJobTests.py` and `AllTests.py` respectively on Windows:
```
Traceback (most recent call last):
  File "FunctionalTests\Jobs\BuildJobTests.py", line 258, in <module>
    job_file_generator = JobFileGenerator(job, job_id=job_id)
  File "FunctionalTests\Jobs\BuildJobTests.py", line 34, in __init__
    self.job.get_default_parameters()
  File "C:\hostedtoolcache\windows\Python\2.7.18\x64\lib\site-packages\mdanse-1.5.0-py2.7-win-amd64.egg\MDANSE\Framework\Configurable.py", line 252, in get_default_parameters
    cfg=REGISTRY["configurator"][typ](name, **kwds)        
KeyError: 'netcdf_input_file'
Traceback (most recent call last):
  File "FunctionalTests\Jobs\AllTests.py", line 37, in <module>
    run_test()
  File "FunctionalTests\Jobs\AllTests.py", line 32, in run_test
    if not unittest.TextTestRunner(verbosity=2).run(suite()).wasSuccessful():
  File "FunctionalTests\Jobs\AllTests.py", line 24, in suite
    modules = [__import__(os.path.splitext(f)[0],globals(),locals(),[],-1) for f in files]
  File "D:\a\MDANSE\MDANSE\Tests\FunctionalTests\Jobs\test_gromacs_trr.py", line 6, in <module>
    import netCDF4
  File "C:\hostedtoolcache\windows\Python\2.7.18\x64\lib\site-packages\netCDF4\__init__.py", line 3, in <module>
    from ._netCDF4 import *
ImportError: DLL load failed: The specified module could not be found.
```
What's puzzling about it is that I have no idea where the key `netcdf_input_file` comes from, and that `_netCDF4` file clearly exists:
```
Directory of C:\hostedtoolcache\windows\Python\2.7.18\x64\Lib\site-packages\netCDF4

05/09/2022  09:17 AM    <DIR>          .
05/09/2022  09:17 AM    <DIR>          ..
12/[21](https://github.com/ISISNeutronMuon/MDANSE/runs/6349010748?check_suite_focus=true#step:8:21)/2018  03:51 PM            35,547 utils.py
12/21/2018  03:51 PM            25,997 utils.pyc
12/21/2018  03:51 PM           973,8[24](https://github.com/ISISNeutronMuon/MDANSE/runs/6349010748?check_suite_focus=true#step:8:24) _netCDF4.pyd
12/21/2018  03:51 PM               686 __init__.py
12/21/2018  03:51 PM               849 __init__.pyc
               5 File(s)      1,036,903 bytes
               2 Dir(s)  94,164,[25](https://github.com/ISISNeutronMuon/MDANSE/runs/6349010748?check_suite_focus=true#step:8:25)0,624 bytes free
```

**Comment by RastislavTuranyi** [comment]
Ok, update;

- It seems the issue with Ubuntu is actually that the netCDF4 library can't really use the netcdf apt package; instead it needs the libraries it itself installs.
- The KeyError happens when it tries to create the test for `StructureFactorFromScatteringFunction`; I have no idea at the moment why it can't find the `NetCDFInputFileConfigurator`

**Comment by eurydice76** [comment]
It can't find the `NetCDFInputFileConfigurator` because that means that there is an error inside that module which prevents the `import` from working well.

**Comment by eurydice76** [comment]
What is strange is that on my machine, nothing wrong happens when loading that configurator.

**Comment by eurydice76** [comment]
However, I still have failing tests for CASTEP for example. Does this test fails also on your machine ?

**Comment by RastislavTuranyi** [comment]
Yes, those tests are failing for me as well, and they are failing on the working MacOS pipeline too. I think it may be something we fixed after this branch was created.

And, yes, it is very weird that the configurator is not initialised, as is the fact that the Ubuntu pipeline is now failing because MMTK can't find `_netcdf`, even though during its installation it says `NetCDF header found --> /opt/hostedtoolcache/Python/2.7.18/x64/include/python2.7/Scientific/netcdf.h`.

**Comment by RastislavTuranyi** [comment]
I see what's going on; the `NetCDFInputFileConfigurator` is failing because it itself import netCDF4, which the pipeline has an issue with; therefore, what remains is to find out why it can't import the `_netCDF4` module which clearly exists.

**Comment by eurydice76** [comment]
Good, you're on the way.

**Comment by RastislavTuranyi** [comment]
Almost everything works now:

- the problem on Ubuntu was that `Scientific.IO.NetCDF` was using the `netcdf.so` library from the netCDF4 installation instead of the apt installation, so all the relevant libraries had to be copied.
- on Windows, netCDF4 couldn't find the netcdf DLLs, so they had to be copied to the netCDF4 directory in site packages.
- when you edited `BuildJobTests.py`, you accidentally overwrote some of the changes implemented in #66; I have essentially merged together the changes from both branches + some extra (netCDF4 implements a context manager!)

The only issue is that the Ubuntu 20 pipeline has a permission problem in one of the tests.

**Comment by eurydice76** [comment]
Well done ! Sorry for the mistake.

**Comment by eurydice76** [comment]
This is very strange that you do not have the right to write on /tmp on the ubuntu 20.04 server. This is exactly the aim of this directory: allowing anybody to write on it. You should set the permission on this directory by  doing a `chmod -R 777 /tmp`

**Comment by eurydice76** [comment]
Can you login on the ubuntu 20.04 machine and send me the return value of `ls -lrth /` ?

**Comment by RastislavTuranyi** [comment]
I have made changes to `BuildJobTests.py` so that the output file path of the MDANSE jobs is directly `../../../Data/Jobs_reference_data` rather than outputting it to temp and then copying it from temp to the reference data folder. This has solved the permission issue.

File: Src/GUI/Plugins/PlotterPlugin.py [hunk 1]
```diff
@@ -472,11 +472,11 @@ def on_load_data(self, event=None):
             basename = baselist[i]    
             filename = filelist[i]
             
-            f = NetCDFFile(filename,"r")
+            f = netCDF4.Dataset(filename,"r")
```
Comment (RastislavTuranyi) [inline Src/GUI/Plugins/PlotterPlugin.py:475]: When/where does this file get closed?

Comment (eurydice76) [reply Src/GUI/Plugins/PlotterPlugin.py:475]: This file should be closed at the end of the method.

File: Src/GUI/Plugins/PlotterPlugin.py [hunk 2]
```diff
@@ -487,7 +487,7 @@ def on_load_data(self, event=None):
                         data[k]['axis'] = []
                 else:
                     data[k]['axis'] = []
-                data[k]['data'] = _vars[k].getValue()
+                data[k]['data'] = _vars[k][:]
                 data[k]['units'] = getattr(_vars[k],"units","au")
             
             unique_name = self.unique(basename, self.plugin._dataDict)
```
Comment (RastislavTuranyi) [inline Src/GUI/Plugins/PlotterPlugin.py:493]: Also, why is `basename` defined inside the for loop, but referenced only outside of it?

Comment (eurydice76) [reply Src/GUI/Plugins/PlotterPlugin.py:493]: No, basename is defined at the `for i in range(len(filelist))` level and use at the end of that block. I do not see any problem here.

File: Src/GUI/Plugins/PlotterPlugin.py (hunk 1 above)
Comment (eurydice76) [reply Src/GUI/Plugins/PlotterPlugin.py:475]: Yeap exactly

**Comment by eurydice76** [comment]
Hi @sanghamitra-mukhopadhyay, could you please review asap this important PR so that we can develop based on this ?

**Review by sanghamitra-mukhopadhyay — APPROVED** [review]

**Comment by sanghamitra-mukhopadhyay** [comment]
I am merging this pull request. @RastislavTuranyi can you please update the compilation documentation to reflect this change including the current draft version of the Users Guide ? Thanks.

**Pull request merged**